soroban-sdk/src/crypto.rs

```diff
@@ -1,5 +1,54 @@
 //! Crypto contains functions for cryptographic functions.
-use crate::{env::internal, unwrap::UnwrapInfallible, Bytes, BytesN, Env};
+
+use crate::{
+    env::internal, unwrap::UnwrapInfallible, Bytes, BytesN, ConversionError, Env, IntoVal,
+    TryFromVal, Val,
+};
+
+/// A wrapper type for a 32-byte cryptographic hash.
+///
+/// This struct is designed to be used in contexts where a hash value generated
+/// by a secure cryptographic function is required.
+pub struct Digest(BytesN<32>);
```

Comment:
I've added it in https://github.com/stellar/stellar-xdr/pull/182 (and associated links to "next", and rs-xdr are in the PR description).

Reply:
I apologise, I think adding this type as an SDK type was a mistake, on my part.

As I look at the other auth types, like the Context, we don't SDK type them. As an input type it doesn't really matter if it is typed or not since auth invocations are special. My concern about the type being used in a return is largely meaningless, and what we should do instead of typing it as "Hash<N>" is to render it into the contract spec as a "BytesN<N>". The reason for this is that an outside implementer can't rely on a legit Hash being returned. If anything, it could be harmful to allow "Hash" to be claimed in any official sense because a contract could just return random data for it's "Hash" and use a spec that misleads.

I realise this will be a little more work, but I think we can remove the spec type we added as a bug fix, and change the mapping logic in this PR to map Hash<N> to BytesN<N> in the spec.

Wdyt?

Reply:
I apologise, I think adding this type as an SDK type was a mistake, on my part.

As I look at the other auth types, like the Context, we don't SDK type them. As an input type it doesn't really matter if it is typed or not since auth invocations are special. My concern about the type being used in a return is largely meaningless, and what we should do instead of typing it as "Hash<N>" is to render it into the contract spec as a "BytesN<N>". The reason for this is that an outside implementer can't rely on a legit Hash being returned. If anything, it could be harmful to allow "Hash" to be claimed in any official sense because a contract could just return random data for it's "Hash" and use a spec that misleads.

I realise this will be a little more work, but I think we can remove the spec type we added as a bug fix, and change the mapping logic in this PR to map Hash<N> to BytesN<N> in the spec.

Wdyt?

Reply:
I've opened this for the fix up:
- https://github.com/stellar/stellar-xdr/pull/184